lib/caqti_connection_sig.mli

```diff
@@ -135,14 +135,16 @@ module type S = sig
     ('b -> 'c -> 'c) ->
     'a -> 'c -> ('c, [> Caqti_error.call_or_retrieve] as 'e) result future
   (** Combining {!call} with {!Response.fold}, this sends a request to the
-      database and folds over the result rows. *)
+      database and folds over the result rows in a single blocking operation.
+  *)
```

Comment:
No, the callback should be non-blocking, in which case the whole call is non-blocking.  The difference from `fold_s` is rather than since the callback does not have access to the concurrency monad, it won't be able to do any IO.

Reply:
I decided to update this comment after I had a look at the code of the postgresql driver where fold is defined as:
```ocaml
let fold f {row_type; result} acc =
  let n = result#ntuples in
  let rec loop i acc =
    if i = n then Ok acc else
      (match decode_row ~uri (result, i) row_type with
       | Ok y -> loop (i + 1) (f y acc)
       | Error _ as r -> r) in
  return (loop 0 acc)
```
which means `loop 0 acc` will be a blocking operation since it doesn't use `Lwt` promises and `bind` operation internally.
At least that's what I understood from Lwt's manual:
> A callback, like the `f` that you might pass to `Lwt.bind`, is an ordinary OCaml function. Lwt just handles ordering calls to these functions
> ...
> This simplifies reasoning about critical sections: all the code in one callback cannot be interrupted by any of the code in another callback. However, it also carries the danger that if a single callback takes a very long time, it will not give Lwt a chance to run your other callbacks.

As far as I understand, for it to be non-blocking it should either use `return` and `bind` internally or be expressed as `fold_s (fun x acc -> return (f x acc))`.

I think it actually makes sense to have a blocking version of fold (and eventually iter) in caqti.

What I meant by "blocking" here is that the fold will be executed at once and cannot be interleaved with other Lwt callbacks. I think it is a significant difference from `fold_s` that required clarification since the whole fold loop could be a lengthy operation.

What do you think?

Reply:
As I understand blocking it means that the whole process will halt waiting for IO.  In the above code for the PostgreSQL driver the IO is already complete, the results being stored in the response.  So, I think it is wrong to call it blocking.  If the callback blocks, then it is, but I would consider passing a blocking function a bug, or at least undesirable when using Lwt or Async.

Reply:
Ok, I'll remove that statement then!

Reply:
Good, otherwise it's ready for merge.